harness/determined/pytorch/_callback.py

```diff
@@ -126,6 +135,16 @@ def on_training_epoch_end(self, epoch_idx: int) -> None:
         """
         pass
 
+    def on_training_workload_end(self, metrics: Dict[str, Any]) -> None:
+        """
+        Run on end of a training workload. Each workload contains a maximum of ``scheduling_unit``
+        batches, as defined in the experiment config.
+
+        .. warning::
+            This callback only executes on the chief GPU when doing distributed training.
```

Comment:
Right now only the chief has access to the `storage_id` that gets passed into this callback, but I guess we could broadcast it to all workers if needed.

However, I was working under the impression that this callback would be used to kick off some external process related to the uploaded checkpoint, in which case you would probably only want a single worker to use the callback, no?  I think that was the most probable use-case we came up with, in which case you probably only want `on_checkpoint_upload_end` being called once.

Reply:
hm?  This comment is about `on_training_workload_end`

Reply:
Oops, misread where you were commenting on. I think we want both of these callbacks to only fire on the chief worker, though, since also for `on_training_workload_end` only the chief has access to the relevant metrics.  We could again broadcast them to all workers, but if the user wants this, they could just do the broadcasting themselves in the callback.

Reply:
actually they _can't_ do the broadcasting themselves, because the master would hang waiting for the broadcast to be received by the workers, while the workers were hanging waiting for the master to finish its `on_train_workload_end`.

> only the chief has access to the relevant metrics
Yeah, I was kind of afraid of this.  Do you think this is going to be a problem for users?

Reply:
Ah right, since `on_train_workload_end` will only currently be called on the chief.

So, I could refactor so that the broadcasting works and all workers have access to the training metrics, but if we think the most-likely use-case for `on_train_workload_end` is to pass the training metrics to some other visualization or logging backend (as indicated by the OSS user who last asked about this), then you would only want one worker to be calling `on_train_workload_end`.

In the spirit of maximum flexibility and minimal assumptions, though, I will try to refactor both `on_train_workload_end` and `on_checkpoint_upload_end` so that they're called on every worker. If the user only wants them to fire on the chief, they can write their callbacks to have a reference to their own `PyTorchTrial` instances, so they can have `if self.trial.is_chief ...` conditionals as needed.

Reply:
I like this, but there's still one more complication, and that's that the training metrics are still just worker 0's metrics, unless the user specifically set `average_training_metrics` to true in their config.

This is a bit of a hot take, but what do you think about making that value true either:
- by by default, and only call `on_training_workload_end` from all workers when it is set to true, or
- always, and always call `on_training_workload_end` with the correct metrics on every worker

Right now, it seems like a bit of a gotcha that the training metrics we broadcast to every worker are just worker 0's metrics by default.  The reason the system is like that right now is that the first dev to write our dtrain support thought training metrics were pretty dumb, and the first customer who wanted our dtrain support literally only rated us on wall clock performance.  (not even model accuracy!)

Reply:
Ugh, yep, totally forgot that `average_training_metrics` is `False` by default.  Thought I had manually tested for this, but I must have have it set to `True` or not tested what I thought I did; my bad.

I actually wrote a Jira ticket a while ago ([DET-7240]) to benchmark the cost of changing the default to `True`, since it seemed strange to me that it wasn't already the default and the cost of making the switch seemed low. So, my opinion is clear.  I haven't managed to revisit the ticket since opening it, though.